### PR TITLE
feat. add a menu item 'expand-select' to expand subtree in search view

### PR DESCRIPTION
**Pull request by pagict**
Implementation of #206032 

image

### CODE REVIEW

**Comment by RedCMD** [comment]
Middle click (scroll click/button 3) is often used for toggling between expanding and collapsing nested items
(just like folding ranges in the editor)

**Comment by pagict** [comment]
@microsoft-github-policy-service agree

File: src/vs/workbench/contrib/search/browser/searchActionsFind.ts [hunk 1]
```diff
@@ -98,6 +98,31 @@ registerAction2(class ExcludeFolderFromSearchAction extends Action2 {
...
+						ContextKeyExpr.or(Constants.SearchContext.FileFocusKey, Constants.SearchContext.FolderFocusKey),
+						Constants.SearchContext.HasSearchResults
+					),
+				}
+			]
+		});
+	}
+
+	override async run(accessor: ServicesAccessor, args: any): Promise<any> {
+		await expandSelectSubtree(accessor);
```
Comment (andreamah) [inline src/vs/workbench/contrib/search/browser/searchActionsFind.ts:122]: no need to await if the function is not async

File: src/vs/workbench/contrib/search/browser/searchActionsFind.ts [hunk 2]
```diff
@@ -335,6 +360,16 @@ function getMultiSelectedSearchResources(viewer: WorkbenchCompressibleObjectTree
 		.filter((renderableMatch): renderableMatch is URI => (renderableMatch !== null));
 }
 
+function expandSelectSubtree(accessor: ServicesAccessor) {
+	const viewService = accessor.get(IViewsService);
+	const searchView = getSearchView(viewService);
+	if (searchView) {
+		const viewer = searchView.getControl();
+		const selected = viewer.getFocus()[0];
+		viewer.expand(selected);
```
Comment (andreamah) [inline src/vs/workbench/contrib/search/browser/searchActionsFind.ts:369]: This only expands one level. The call will need to be such that recursive is true.

**Review by andreamah — CHANGES_REQUESTED** [review]
Just a few small changes 👍

**Comment by andreamah** [comment]
> Middle click (scroll click/button 3) is often used for toggling between expanding and collapsing nested items (just like folding ranges in the editor)

I think the search view already supports this, where middle click expands one level. Is this what you meant, or is there still missing functionality?

**Comment by pagict** [comment]
> > Middle click (scroll click/button 3) is often used for toggling between expanding and collapsing nested items (just like folding ranges in the editor)
> 
> I think the search view already supports this, where middle click expands one level. Is this what you meant, or is there still missing functionality?

I meant for recursive expansion. I will update my code to make work, thx for reviewing.

**Comment by RedCMD** [comment]
> I think the search view already supports this, where middle click expands one level. Is this what you meant, or is there still missing functionality?

yeah I meant for recursively nested items
The same as Folding ranges in the editor

**Comment by andreamah** [comment]
> > I think the search view already supports this, where middle click expands one level. Is this what you meant, or is there still missing functionality?
> 
> yeah I meant for recursively nested items The same as Folding ranges in the editor

As in the middle click should recursively open nested items? Does this only happen in the editor with nested ranges, or does it happen in other tree views?

**Comment by RedCMD** [comment]
yes, I would like middle click to recursively open nested items.
I think it only works for Folding ranges atm
I don't think any other tree view supports it

I'm pretty sure there was something else that did support it
but I can't remember if it was even in VSCode or not

**Comment by pagict** [comment]
I tried and also checked the code, "middle mouse for recursively expansion" only works for foldings in the editor. 

https://github.com/microsoft/vscode/blob/1f94e5cd54ce0a7bc503a3f95a3742ddc5980151/src/vs/editor/contrib/folding/browser/folding.ts#L492-L499

**Comment by andreamah** [comment]
Do we use the language "Expand Selected" anywhere else? Otherwise, I would prefer just "Expand" since it's implied that it's acting on selected elements.

**Comment by pagict** [comment]
> Do we use the language "Expand Selected" anywhere else? Otherwise, I would prefer just "Expand" since it's implied that it's acting on selected elements.

no, the term "Expand Selected" does not in anywhere else. should I modify it as "Expand the Tree" or something similar, since this is quite a different behavior to what "Expand" implies.

**Comment by andreamah** [comment]
image

The button that expands/collapses all (in the top toolbar) uses the term "Expand all" to expand all of the tree nodes, so I would expect that right-clicking and "expanding" a node would recursively open the node.

**Comment by pagict** [comment]
> image
> 
> The button that expands/collapses all (in the top toolbar) uses the term "Expand all" to expand all of the tree nodes, so I would expect that right-clicking and "expanding" a node would recursively open the node.

update as "Expand All", please kindly review

**Comment by pagict** [comment]
I think they both are good, and how about "Expand from This"? or maybe should be more succinct.

I don't have any preferences on this

**Comment by andreamah** [comment]
I think "Expand Recursively" would be the best from my perspective, since it seems less wordy in my opinion.

**Review by andreamah — APPROVED** [review]
Thanks!

**Review by amunger — APPROVED** [review]

**Pull request merged**